consensus/propagation/catchup.go

```diff
@@ -1,54 +1,108 @@
+//
+// todo: add a request limit for each part to avoid downloading the block too
+// many times. atm, this code will request the same part from every peer.
+func (blockProp *Reactor) retryWants(currentHeight int64, currentRound int32) {
+	data := blockProp.dumpAll()
+	peers := blockProp.getPeers()
+	for _, prop := range data {
+		height, round := prop.compactBlock.Proposal.Height, prop.compactBlock.Proposal.Round
 
-// requestMissingBlocks is called when a node is catching up and needs to
-// request all previous blocks from the compact block source peer.
-func (blockProp *Reactor) requestMissingBlocks(targetHeight int64, targetRound int32) {
-	// if we're missing heights, we will be requesting them from the connected peers
-	for height := blockProp.currentHeight + 1; height < targetHeight; height++ {
-		heightProposal, partSet, found := blockProp.GetProposal(height, -1)
-		if !found {
-			blockProp.Logger.Error("couldn't find proposal", "height", height, "round", -1)
+		if height == currentHeight && round == currentRound {
 			continue
 		}
-		if partSet.BitArray() == nil {
-			panic("partSet is nil when getting proposal!")
-		}
-		missingParts := partSet.BitArray().Not()
-		wantPart := &proptypes.WantParts{
-			Parts:  missingParts,
-			Height: height,
-			Round:  heightProposal.Round,
-		}
-		blockProp.broadcastWants(wantPart, blockProp.self)
-	}
 
-	// if we're missing rounds for the current height, request them from all the peers
-	for round := int32(0); round < targetRound; round++ {
-		_, partSet, found := blockProp.GetProposal(targetHeight, round)
-		if !found {
-			blockProp.Logger.Error("couldn't find proposal", "height", targetHeight, "round", round)
+		if prop.block.IsComplete() {
 			continue
 		}
-		if partSet.BitArray() == nil {
-			panic("partSet is nil when getting proposal!")
+
+		// only re-request original parts that are missing, not parity parts.
+		missing := prop.block.BitArray().Not()
+		if missing.IsEmpty() {
+			// this should never be hit due to the check above.
+			continue
 		}
-		missingParts := partSet.BitArray().Not()
-		wantPart := &proptypes.WantParts{
-			Parts:  missingParts,
-			Height: targetHeight,
-			Round:  round,
+
+		// make requests from different peers
+		peers = shuffle(peers)
+
+		for _, peer := range peers {
+			mc := missing.Copy()
+			reqs, has := peer.GetRequests(height, round)
+			if has {
+				mc = mc.Sub(reqs)
+			}
+
+			if mc.IsEmpty() {
+				continue
+			}
+
+			e := p2p.Envelope{
+				ChannelID: WantChannel,
+				Message: &protoprop.WantParts{
+					Parts:  *missing.ToProto(),
+					Height: height,
+					Round:  round,
+					Prove:  true,
+				},
+			}
+
+			if !p2p.TrySendEnvelopeShim(peer.peer, e, blockProp.Logger) { //nolint:staticcheck
+				blockProp.Logger.Error("failed to send want part", "peer", peer, "height", height, "round", round)
+				continue
+			}
+
+			// keep track of which requests we've made this attempt.
+			missing.Sub(mc)
+			peer.AddRequests(height, round, missing)
```

Comment:
note that need to test this function in a unit test and ensure that we're only actually max requesting parts once in a follow up